### PR TITLE
Make sparse tensor types extend the existing tensor types

### LINKED ISSUE #142: Unify sparse and dense tensor class design
The sparse tensors should use exactly the same interfaces/types (and much of the same code) as the dense tensors, since we're supposed to be modeling high-level tensors with these interfaces, and **not** their implementation details.  Consider this: in what world is a sparse tensor not a tensor?  (Answer: Theano.)

To start, the mixin `theano.sparse.basic._sparse_py_operators` needs to be replaced with a possibly generalized version of `theano.tensor.basic._tensor_py_operators`.

Next, `theano.sparse.basic.SparseVariable` should extend `theano.tensor.var.TensorVariable` (again, this class could be generalized a little, if necessary).  Same with `theano.sparse.type.SparseType` and `theano.tensor.type.TensorType`.

Overall, we should make the changes necessary to allow the use of fundamental `Op`s with sparse tensor arguments (e.g. `Op`s like `*Subtensor*`).  Speaking of `*Subtensor*`, all the `theano.sparse.basic.GetItem*` and `theano.sparse.basic.Transpose` `Op`s are painfully limited and redundant when `*Subtensor*` is compatible with sparse tensors.  None of those `Op`s actually do anything specific to sparse tensors, aside from some `csr` and `csc` argument conversions that could be done almost anywhere (if at all).

_Originally posted by @brandonwillard in https://github.com/pymc-devs/Theano-PyMC/issues/52#issuecomment-720233156_

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR refactors the sparse tensor type structure so that sparse types extend the existing tensor types.  A new meta type is used to construct a `DenseTensorType` that can be used to distinguish dense from sparse `TensorType`s.  

Closes #142.

- [x] Create a `DenseTensorVariable` (using the same approach as `DenseTensorType`)
- [x] Update more optimizations so that they exclude `SparseTensor`s
- [ ] Convert sparse to dense when used with some `Op`s (e.g. `Elemwise`, `CAReduce`)

### CODE REVIEW

File: aesara/sparse/basic.py [hunk 1]
```diff
@@ -22,6 +22,9 @@
 from aesara.misc.safe_asarray import _asarray
 from aesara.sparse.type import SparseType, _is_sparse
 from aesara.sparse.utils import hash_from_sparse
+
+# TODO:
```
Comment (twiecki) [inline aesara/sparse/basic.py:26]: For when is this TODO?

File: aesara/sparse/basic.py [hunk 2]
```diff
@@ -123,8 +127,9 @@ def _is_dense(x):
     return isinstance(x, np.ndarray)
 
 
-# Wrapper type
-def as_sparse_variable(x, name=None):
+# TODO:
```
Comment (twiecki) [inline aesara/sparse/basic.py:130]: Maybe link to the GH issue.

**Review by twiecki — APPROVED** [review]

File: aesara/sparse/basic.py [hunk 3]
```diff
@@ -355,8 +360,7 @@ def __getitem__(self, args):
         return ret
 
 
-class SparseVariable(_sparse_py_operators, Variable):
-    dtype = property(lambda self: self.type.dtype)
+class SparseVariable(_sparse_py_operators, TensorVariable):
```
Comment (michaelosthege) [inline aesara/sparse/basic.py:363]: Why are the supertypes in a different order compared to the `SparseConstant` below?

File: aesara/sparse/basic.py [hunk 4]
```diff
@@ -413,6 +416,12 @@ def __repr__(self):
 SparseType.Variable = SparseVariable
 SparseType.Constant = SparseConstant
 
+# TODO:
+# @_as_tensor_variable.register(SparseVariable)
+# @_as_tensor_variable.register(SparseConstant)
+# def _as_tensor_sparse(x, name, ndim):
+#     return x
+
```
Comment (michaelosthege) [inline aesara/sparse/basic.py:424]: This TODO could also use an explanation or link to a Github issue

File: aesara/sparse/basic.py [hunk 5]
```diff
@@ -442,7 +451,7 @@ def bsr_matrix(name=None, dtype=None):
 csr_fmatrix = SparseType(format="csr", dtype="float32")
 bsr_fmatrix = SparseType(format="bsr", dtype="float32")
 
-all_dtypes = SparseType.dtype_set
+all_dtypes = list(SparseType.dtype_specs_map.keys())
```
Comment (michaelosthege) [inline aesara/sparse/basic.py:454]: Did you make it a list so it can be appended at runtime?

File: aesara/sparse/type.py [hunk 6]
```diff
@@ -60,48 +60,52 @@ class SparseType(Type):
...
             raise Exception(
-                "You can't make SparseType object as SciPy" " is not available."
-            )
-        dtype = str(dtype)
-        if dtype in self.dtype_set:
-            self.dtype = dtype
-        else:
-            raise NotImplementedError(
-                f'unsupported dtype "{dtype}" not in list', list(self.dtype_set)
+                "You can't make SparseType object when SciPy is not available."
```
Comment (michaelosthege) [inline aesara/sparse/type.py:86]: Is it even a supported use case to NOT have scipy?

File: aesara/tensor/basic.py [hunk 7]
```diff
@@ -341,8 +341,13 @@ def get_scalar_constant_value(
 
             if isinstance(data, np.ndarray):
                 return numpy_scalar(data).copy()
-            else:
-                return data
+
+            from aesara.sparse.type import SparseType
```
Comment (michaelosthege) [inline aesara/tensor/basic.py:345]: Could a whitelist of supported types be an alternative here?
After all that import is sort of cyclic...

File: tests/sparse/test_basic.py [hunk 8]
```diff
@@ -1017,22 +1018,45 @@ class TestConversion:
...
+        assert isinstance(res, SparseConstant)
+
+        a = sp.sparse.csr_matrix(test_val)
+        s = as_sparse_or_tensor_variable(a)
+        from aesara.tensor.exceptions import NotScalarConstantError
+
+        with pytest.raises(NotScalarConstantError):
+            aet.get_scalar_constant_value(s, only_process_constants=True)
+
+    # TODO:
```
Comment (michaelosthege) [inline tests/sparse/test_basic.py:1048]: What does this test depend on? Could it become XFAIL instead or link to an issue?

File: tests/sparse/test_basic.py [hunk 9]
```diff
@@ -1607,6 +1631,33 @@ def test_int32_dtype(self):
...
+        op_types = set(type(n.op) for n in applys_between([x, y], [res]))
+        assert sparse.basic.StructuredDot in op_types
+        assert aet.math.Dot not in op_types
+
+        res = aet.dot(aet.second(1, x), y)
+        op_types = set(type(n.op) for n in applys_between([x, y], [res]))
+        assert sparse.basic.StructuredDot in op_types
+        assert aet.math.Dot not in op_types
+
+    # @aesara.config.change_flags(compute_test_value="raise", optimizer_verbose=True)
```
Comment (michaelosthege) [inline tests/sparse/test_basic.py:1660]: Looks like this `change_flags` is not needed?

File: aesara/sparse/basic.py (hunk 3 above)
Comment (brandonwillard) [reply aesara/sparse/basic.py:363]: Not sure.

File: aesara/sparse/basic.py (hunk 4 above)
Comment (brandonwillard) [reply aesara/sparse/basic.py:424]: I was waiting to merge #302.

File: aesara/sparse/type.py (hunk 6 above)
Comment (brandonwillard) [reply aesara/sparse/type.py:86]: That's how this project was originally designed, I believe, but there's no point in supporting that optionality now.

File: tests/sparse/test_basic.py (hunk 9 above)
Comment (brandonwillard) [reply tests/sparse/test_basic.py:1660]: No, it's not; that was left over from debugging.

File: aesara/sparse/basic.py (hunk 1 above)
Comment (brandonwillard) [reply aesara/sparse/basic.py:26]: I was waiting to merge #302.

File: aesara/sparse/basic.py (hunk 2 above)
Comment (brandonwillard) [reply aesara/sparse/basic.py:130]: I was waiting to merge #302.

**Comment by brandonwillard** [comment]
This sparse implementation only supports matrices, but we could cover more ground with [`sparse`](https://sparse.pydata.org/en/stable/index.html).  It's an additional dependency, but seems like it could be worth it.

File: aesara/tensor/basic.py (hunk 7 above)
Comment (brandonwillard) [reply aesara/tensor/basic.py:345]: Given the nature of this function (i.e. `get_scalar_constant_value`), it should also be converted to a dispatch function.  That way, we can keep irrelevant types out of `aesara.tensor.basic` entirely (e.g. `Assert`).

Comment (brandonwillard) [reply aesara/tensor/basic.py:345]: Actually, we have an issue for this: #96.

**Pull request closed without merging**